E2E: Add browser context instance separation

WunderBart opened the pull request:
## Proposed changes

1. **Introduce a global browser instance with multiple context support.**

   With the addition of `newContext` and `newPage` proxy traps that allow us to pass our default context options, we can now simply call e.g. `await browser.newPage()` to create a fresh new page within our tests as many times as we want - no need to use the `setupHook` anymore! All those pages and contexts that we create within our tests will also be handled in the `teardown` hook and, if an error occurred, will have dedicated artifacts created and saved. The biggest advantage here is that the direct access to separate contexts allows us to authenticate multiple accounts per test (also in parallel!) without the need to clear cookies, local storage, etc. Having a 100% separation should be a boost to the confidence in our e2e tests.

2. **Move Jest/Playwright environment setup to the `calypso-e2e` package.**

   All the files related to setting up Jest with Playwright are now within the dedicated package - a single source of truth for running our e2e tests.

4. **Create `envVariables` module for easier reference of the Node env variables.**

   The `envVariables` module returns a strongly typed object with environment variables we use in our e2e tests. All those variables are also casted to their target types and validated at runtime so we don't need to do any extra handling outside that module. For extra safety against unwanted side fx, the `envVariables` Are exported as a readonly object. I think the big value here is also having a single source of true for all the env vars.

5. **Rename `SAVE_AUTH_COOKIES` to `AUTHENTICATE_ACCOUNTS`.**
   
   Instead of passing true/false value and defining the commonly used accounts to pre-authenticate within the global setup, we can now pass those accounts directly via the `AUTHENTICATE_ACCOUNTS` variable for better visibility and control.

## Testing instructions

- All tests should pass.
- Artifacts should be saved as expected (log, trace, screenshot, and video).
- Pre-authenticating with `AUTHENTICATE_ACCOUNTS` should work as expected (described above).

Fixes https://github.com/Automattic/wp-calypso/issues/56805 and https://github.com/Automattic/wp-calypso/issues/56449.

WunderBart committed:
Add handling for multiple contexts and pages

WunderBart committed:
Proxy browser instance to bind custom context options

WunderBart committed:
Add env-variables.js helper

WunderBart committed:
Move e2e jest config to calypso-e2e package

WunderBart committed:
Update calypso-e2e package to use new env config

WunderBart committed:
Update tests/e2e to use new env from calypso-e2e

WunderBart committed:
Do some renaming

WunderBart committed:
Validate env variables at runtime

WunderBart committed:
Create new context for each user in describe.each blocks

WunderBart committed:
Fix failing suites

WunderBart committed:
Fix env var check

WunderBart committed:
Remove accidental log

WunderBart committed:
Move logger config to env setup

WunderBart committed:
Remove unnecessary validation

WunderBart committed:
Record videos only for actual tests

WunderBart committed:
Restore screenshots path

WunderBart committed:
Update post like spec to use separate contexts

As per https://github.com/Automattic/wp-calypso/pull/59074#discussion_r783946783

WunderBart committed:
Fix env vars reference in i18n specs

WunderBart committed:
Fix deleted BrowserHelper references

WunderBart committed:
Not sure how this one didn't fail :D

WunderBart committed:
Rename SAVE_AUTH_COOKIES to AUTHENTICATE_ACCOUNTS and update docs

WunderBart committed:
Add todo

fullofcaffeine commented on packages/calypso-e2e/src/jest-playwright-config/environment.ts:
```diff
@@ -0,0 +1,198 @@
+
+		// Proxy our E2E browser to bind our custom default context options.
+		// We need to it this way because we're not using the first-party Playwright
+		// and we can't use the playwright.config.js file.
```
Hm, interesting! Why aren't we using the first-party Playwright lib and what are we using instead?

fullofcaffeine replied:
Ah, `jest-playwright`, I assume.

worldomonation replied:
Currently, we use vanilla Jest with some plugins and custom reporters.

Sergio and I had this discussion when he was still with a8c and IIRC the gist was that:
- Jest supports custom reporters while Playwright Test Runner does not;
- We can use other plugins with Jest (eg. Jest Runner Groups) whereas Playwright Test Runner won't.

I do have an item in the backlog, for a 20% time project, to play with the Playwright Test Runner. Its advantages are:
- first-party support for managing multiple pages/contexts and artifacts (screenshot, video) for those.
- interfaces well with simulated devices that Playwright ships with.
- much easier to set up cross-browser testing.

fullofcaffeine replied:
Thanks a lot for elaborating :)

fullofcaffeine commented on test/e2e/docs/test_environment.md:
````diff
@@ -29,7 +29,7 @@ export CONFIG_KEY=<decryption_key_from_a8c_store>
 
 ```
-export SAVE_AUTH_COOKIES=<true/false>
+export AUTHENTICATE_ACCOUNTS=p2User,i18nUser
````
Neat. Do we have any tests that actually use this pre-authentication feature for multiple accounts?

WunderBart replied:
The accounts we pre-authenticate have been set to `simpleSitePersonalPlanUser`, `eCommerceUser`, and `defaultUser`. Before this PR, they were defined in `globalSetup.js`, and with this change, they're going to be passed with this env var for better visibility. So it's not a new feature, just an existing one that got a bit refactored.

fullofcaffeine commented on packages/calypso-e2e/src/jest-playwright-config/environment.ts:
```diff
@@ -0,0 +1,198 @@
+	return text.replace( /[^a-z0-9]/gi, '-' ).toLowerCase();
+};
+
+class JestEnvironmentPlaywright extends JestEnvironmentNode {
```
Nit: might worth adding some basic jsdoc describing this and mentioning where this is [used from](https://github.com/Automattic/wp-calypso/pull/59811/files#diff-046bb2a4070f4d7e0d5ec74a917e2041b5d0d97a3b8498d1f2ae846543f80ae9R8) from those not too familiar with the Jest API.

WunderBart commented:
Great questions, @fullofcaffeine! 🙌 

> 1. As you mention, we already called `BrowserContext`'s `newPage()` from the [setupHooks](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/hooks.ts#L53) method. We also had what seemed to be another implementation of `newPage()` [in the browser-manager module](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/browser-manager.ts#L57). Do you mind sharing a bit about the historical reasons, how they were used and why we had those? Maybe that was part of the reason we refactored into this version?

Historically, I believe this structure was chosen as it was how we managed the browser/page creation with Selenium. This made more sense back then, as we had only one context to work with and we had to manually "reset" its instance (e.g. by clearing cookies) if we wanted e.g. another account to authenticate within the same suite (see e.g. the [`wp-support__home-spec.ts`](https://github.com/Automattic/wp-calypso/blob/3c65b726b7cc935afc8e3d0b7a866e2f17a698a3/test/e2e/specs/specs-playwright/wp-support__home-spec.ts)). With Playwright, we get OOTB a very efficient way of doing that (new context is created within milliseconds) with a 100% assurance every new context is fully isolated from the other. Another great thing is that new contexts can now also be created in parallel within a single suite, which makes possible e.g. using multiple accounts at a time that "talk" to each other. This means that, for example in the [wp-invite__new-user.ts](https://github.com/Automattic/wp-calypso/blob/622c4058d0d1576da682d8906186e4fb727abceb/test/e2e/specs/specs-playwright/wp-invite__new-user.ts) suite, we wouldn't need to sign out and back in between the inviter and invitee accounts! This leads me to another limitation that is gone now, and was present in the previous implementation: Because of the single/global context/page principle provided by the `setupHook()`, even if we created a new page within the test (which was possible via the `BrowserHelper.newPage()` method you mentioned), we wouldn't get any failure artifacts for that page **unless** we've explicitly overwritten the global one. Overwriting global/parent params can have all sorts of side effects, which could negatively impact our E2E tests, which are a brittle thing anyway.


> You already mention some advantages of the new implementation in the PR description, and those sound great. It's not clear though if there are any tests that will take advantage of this now. Since this PR doesn't link to a project thread or issue, are there any discussions or P2 posts that discuss this before we eventually decided to go this route?

The direct advantage here is for the suites where we use multiple accounts (e.g. testing a feature on both Simple, and Atomic sites) or simply want a clean state between tests (e.g. Editor block tests), etc. I think the most important thing though is making the implementation itself impose good practices, where the only global thing is the browser, and we have to create contexts and pages explicitly. This way is also less boilerplate and, surprisingly, pretty much vanilla Jest/Playwright, which is great for new devs - no setup hooks, browser managers, etc.

As for the missing discussion / P2 posts - that's totally on me, and I'm sorry for that. The reason I haven't done that is that I wasn't even sure I'll be able to finish this before my parental leave. I was sure I'm going in the right direction here though, but that doesn't really justify not discussing it beforehand. Having said that, I'm of course open to changes, or even rejecting the whole thing if my ideas are considered to be totally missed. I won't be able to respond to your feedback anymore though, as today is my last day before my leave, so I hope I've explained my reasoning well. FYI, I've already spoken to @worldomonation about taking this one over, so the decision is now on his side! 🙇

WunderBart committed:
Add some comments

fullofcaffeine commented on packages/calypso-e2e/src/jest-playwright-config/environment.ts:
```diff
@@ -13,6 +12,13 @@ const sanitizeString = ( text: string ) => {
+/**
+ * This is the place where we make Playwright work with Jest (jest-circus): From
+ * setting up our global Browser instance to handling failure and teardown
+ * hooks.
```
Thanks! <3

worldomonation commented:
I'm taking over this PR now that Bart is on leave.

worldomonation commented:
@fullofcaffeine Bart has explained in detail the first part of the question, so I will focus on the second portion:

>     2. You already mention some advantages of the new implementation in the PR description, and those sound great. It's not clear though if there are any tests that will take advantage of this **now**. Since this PR doesn't link to a project thread or issue, are there any discussions or P2 posts that discuss this before we eventually decided to go this route?

[This task](https://github.com/Automattic/wp-calypso/issues/56805) that I filed back in October 2021 is the basis for the chain of PRs that Bart has been working on, including this one.

As he has explained in https://github.com/Automattic/wp-calypso/pull/59811#issuecomment-1013161320, we've had issues where non-default Pages and BrowserContexts do not save artifacts that are useful for debugging (eg. screenshot, video). This is because our pre- and post-test hooks were working off assumption that only one Page and BrowserContext existed at any given time.

Naturally, this leads to issues when two or more accounts need to interact in a test.

Aside from the example Bart mentioned (with `Invite: New User`), there are numerous other specs that I rewrote to use only single Page and BrowserContext such as `Notifications` and `Reader: Comment` specs. 

Currently, these specs manually clear the authenticated state of the given Page/BrowserContext and proceed to re-authenticate as the intended user. The process of clearing the authenticated state is error-prone as not all state is truly cleared by this action. For some time, the `Notifications` spec would experience relatively high levels of intermittent failure due to this very phenomenon.

This PR, while massive, aims to fully take advantage of the Pages and BrowserContext separation offered by Playwright out of the box _and_ enable the proper screenshot and video capture of non-default Pages and BrowserContexts. This would allow for much more complex test specs in the future.

worldomonation commented:
@dpasque This is a big PR, but a major change to the underlying framework and I felt you should have a heads up.

worldomonation commented:
Ideally I'd like us to come to a decision on whether to make this change or not before this PR goes stale (which is highly likely due to the sheer number of files it touches).

worldomonation reviewed (approved):
Other than a few naming scheme nitpicks, I think this change looks great.

dpasque commented on packages/calypso-e2e/src/env-variables.ts:
```diff
@@ -0,0 +1,124 @@
+	ARTIFACTS_PATH: path.join( process.cwd(), 'results' ),
+};
+
+const castKnownEnvVariable = ( name: string, value: string ): EnvVariableValue => {
```
This dynamic way of ensuring typing is very clever! Nicely done!!

dpasque commented on packages/calypso-e2e/src/env-variables.ts:
```diff
@@ -0,0 +1,124 @@
+	}
+} );
+
+// TODO: Should we add proxy trap to fallback to process.env values?
```
I don't think so at this point! I like providing and explicit, curated list of variables

dpasque commented on packages/calypso-e2e/src/jest-playwright-config/environment.ts:
```diff
@@ -0,0 +1,210 @@
+				if ( this.failure?.type === 'test' ) {
+					event.test.mode = 'skip';
+				}
+				// If a hook has failed, mark all subsequent test steps as failed.
+				// Handling is different compared to test steps because Jest treats
+				// failed hooks differently from tests.
+				if ( this.failure?.type === 'hook' ) {
+					// event.test.mode = 'fail';
+					// This is now commented out as it throws the "Type '"fail"' is not
+					// assignable to type 'BlockMode'" TS error. The accepted options are
+					// "skip", "todo", and "only".
+				}
```
This is default behavior from what I can tell --- should we just take this out for now?

worldomonation replied:
Agreed.

dpasque commented on packages/calypso-e2e/src/jest-playwright-config/environment.ts:
```diff
@@ -0,0 +1,210 @@
+import { Browser, BrowserContext, BrowserContextOptions, chromium, Page } from 'playwright';
+import env from '../env-variables';
+import config from './playwright-config';
+import type { Config, Circus } from 'jest-environment-node/node_modules/@jest/types';
```
This doesn't quite seem like the right import location.... Do we not have `@jest/types` directly that we can import here? I just don't think we want to target the `node_modules` of another package

worldomonation replied:
I've changed this line to import from `@jest/types` directly. Admittedly, I do completely understand Bart's original intentions here, but I think what you describe makes sense.

If for whatever reason the import from `@jest/types` becomes an issue, we can always fall back on the way it was originally implemented.

dpasque commented on packages/calypso-e2e/src/jest-playwright-config/environment.ts:
```diff
@@ -0,0 +1,210 @@
+			case 'test_fn_failure': {
+				this.failure = { type: 'test', name: event.test.name };
+				break;
+			}
```
Right now, because we are only concerned with skipping tests, there are two edge cases that can account for strange behaviot:

1. If a hook fails in a given block, we still try to run later describe blocks in the test. We probably don't want to do that, as those aren't going to be valid if none of the prior tests ran.

2. If a test fails, we will skip all future tests in future describe blocks, but we still run the _hooks_ in those future describe blocks. This can result in cases where we open new pages (see the liking post spec), but do nothing with them. This then muddies the artifacts, as you have videos and screenshots of just blank white screens of those secondary pages that were opened and immediately abandoned.

We can fix both of these by adding a check for failures at the start of each describe block:

```typescript
	case 'run_describe_start': {
		if ( this.failure ) {
			event.describeBlock.mode = 'skip';
		}
		break;
	}
```

worldomonation replied:
@dpasque I added this and triggered an intentional failure in the hook of a spec file, but the behavior observed does not change:

```
    ✕ Go to new post page
    ✕ Enter post title
    ✕ Image block: upload image file
    ✕ Image block: upload image file with reserved URL characters
    ✕ Audio block: upload audio file
    ✕ File block: upload audio file
    ✕ Publish and visit post
    ✕ Confirm Image block is visible in published post
    ✕ Confirm Image block is visible in published post (reserved name)
    ✕ Confirm Audio block is visible in published post
    ✕ Confirm File block is visible in published post
```

To be clear, with this change applied, the expectation is that all steps are marked `Skipped` instead, correct?

worldomonation replied:
I poked around a bit more, and I think I understand what you mean now - I was misunderstanding your comment fundamentally!

With the suggested changes applied, this is the result of `Likes: Post`:

```
    As the posting user
      ✕ Go to the new post page
      ✕ Enter post title
      ✕ Enter post text
      ✕ Publish and visit post
      ✕ Like post
      ✕ Unlike post
    As the liking user
      ○ skipped Go to the published post page
      ○ skipped Login via popup to like the post
```

dpasque replied:
Yes! What you have there in the last comment is what I was thinking!

dpasque reviewed (approved):
This is just awesome!! I love this!

I have a few comments, most small, but probably worth fixing in this PR before merge. But approving because overall this looks and works great!

worldomonation committed:
Revise based on PR review comments:
- remove dead section of code in `test_start`.
- add new case `run_describe_start` to check for failures in previous blocks in order to skip all subsequent describe blocks.
- revise import of '@jest/types';

Commit:
consolidate jest environment and global typings

Commit:
Match jest environment versions for node and jsdom
